src/turbomind/utils/anomaly_handler.cu

```diff
@@ -0,0 +1,387 @@
+            }
+
+            FT_CHECK(0 <= fallback_);
+            FT_CHECK(fallback_ < vocab_size);
+
+            TM_LOG_WARNING("[AnomalyHandler] max_batch_size: %d", max_batch_size);
+            TM_LOG_WARNING("[AnomalyHandler] vocab_size: %d", vocab_size);
+        }
+    }
+
+    void Summarize(std::function<void(const int*, int)> handler)
+    {
+        if (g_level) {
+            check_cuda_error(cudaMemcpyAsync(h_count_.data(),
+                                             d_count_.data().get(),
+                                             sizeof(size_type) * info_.size() * 2,
+                                             cudaMemcpyDefault,
+                                             stream_));
+
+            check_cuda_error(cudaMemcpyAsync(h_is_anomaly_.data(),
+                                             d_is_anomaly_.data().get(),
+                                             sizeof(int) * batch_size_,
+                                             cudaMemcpyDefault,
+                                             stream_));
+
+            check_cuda_error(cudaStreamSynchronize(stream_));
+
+            for (size_t i = 0; i < info_.size(); ++i) {
+                const auto& n_inf = h_count_[i * 2];
+                const auto& n_nan = h_count_[i * 2 + 1];
+                if (n_inf || n_nan) {
+                    TM_LOG_WARNING("[AnomalyHandler][rank=%d] (%s) INF: %s, NaN: %s",
+                                   rank_,
+                                   info_[i].c_str(),
+                                   std::to_string(n_inf).c_str(),
+                                   std::to_string(n_nan).c_str());
+                }
+            }
+
+            handler(h_is_anomaly_.data(), batch_size_);
+        }
+    }
+
+    void Reset()
+    {
+        if (g_level) {
+            if (!info_.empty()) {
+                std::fill_n(h_count_.data(), info_.size() * 2, 0);
+                check_cuda_error(
+                    cudaMemsetAsync(d_count_.data().get(), 0, sizeof(size_type) * info_.size() * 2, stream_));
+                info_.clear();
+            }
+
+            if (batch_size_) {
+                std::fill_n(h_is_anomaly_.data(), batch_size_, 0);
+                check_cuda_error(cudaMemsetAsync(d_is_anomaly_.data().get(), 0, sizeof(int) * batch_size_, stream_));
+                batch_size_ = 0;
+            }
+        }
+    }
+
+    template<class T>
+    void invokeCountAndFixAnomaly(T* data, int64_t size, const std::string& key, int level)
+    {
+        if (g_level && level <= g_level) {
+            FT_CHECK(size >= 0);
+
+            constexpr int block = 512;
+            const int     grid  = (size + block - 1) / block;
+
+            auto idx = info_.size();
+            auto ptr = d_count_.data().get() + idx * 2;
+
+            info_.push_back(key);
+
+            FT_CHECK(info_.size() <= max_entries);
+
+            CountAndFixAnormaly<T, block><<<grid, block, 0, stream_>>>(data,  //
```

Comment:
why do we need `block`  as the template argument?

Reply:
Here `block` is a compile-time constant while `blockDim` in device code is a runtime variable. `cub::BlockReduce` needs a compile-time constant as it's template parameter.